include/envoy/event/timer.h

```diff
@@ -59,11 +60,43 @@ class TimeSystem : public TimeSource {
 public:
   virtual ~TimeSystem() {}
 
+  using Duration = MonotonicTime::duration;
+
   /**
    * Creates a timer factory. This indirection enables thread-local timer-queue management,
    * so servers can have a separate timer-factory in each thread.
    */
   virtual SchedulerPtr createScheduler(Libevent::BasePtr&) PURE;
+
+  /**
+   * Advances time forward by the specified duration, running any timers
+   * along the way that have been scheduled to fire.
+   *
+   * @param duration The amount of time to sleep.
+   */
+  virtual void sleep(const Duration& duration) PURE;
```

Comment:
Q: It seems like the new functions are only used in test code. Is there any way to have this portion of the interface only be in the test code somehow? I haven't been tracking all of these reviews so I'm not sure. (In general I'm a huge fan of this work and think this is awesome.)

Reply:
Thanks and great question!

We could do that. Right now RealTimeSystem is used in production and in also in most tests. IMO we should mostly use SimulatedTimeSystem in tests, but I think there are cases where RealTime is needed, and we'll need the sleep() calls in the integration test framework to actually sleep.

However, I already have a separate include-file and wrapper class for RealTime used in tests (test/test_common/test_time.h). There could be an interface derived from TimeSystem that exposes the sleep() interface, and is implemented by  a new class TestRealTimeSystem and also by SimulatedTimeSystem. Whereas production code would see only TimeSystem and not have a sleep() call.

Would it be cool to do that as a follow-up? It might be a boring long PR with lots of trivial edits. No problem to do it, but  I wouldn't want to mix it with something semantically interesting like this.